curator-framework/src/main/java/org/apache/curator/framework/imps/Watching.java

```diff
@@ -48,15 +53,30 @@
     {
         watcher = null;
         watched = false;
+        curatorWatcher = null;
     }
 
     Watcher getWatcher(CuratorFrameworkImpl client, String unfixedPath)
     {
-        if ( (watcher != null) && (client.getWatcherRemovalManager() != null) )
+        NamespaceWatcher namespaceWatcher = null;
+        if ( watcher != null )
+        {
+            namespaceWatcher = new NamespaceWatcher(client, this.watcher, unfixedPath);
+        }
+        else if ( curatorWatcher != null )
         {
-            return client.getWatcherRemovalManager().add(unfixedPath, watcher);
+            namespaceWatcher = new NamespaceWatcher(client, curatorWatcher, unfixedPath);
         }
-        return watcher;
+
+        if ( namespaceWatcher != null )
+        {
+            if ( client.getWatcherRemovalManager() != null )
+            {
+                client.getWatcherRemovalManager().add(namespaceWatcher);
+            }
+        }
+
+        return namespaceWatcher;
     }
 
     boolean isWatched()
```

Comment:
unrelated to PR, but a name like `needsImplicitWatch` might better explain the purpose